network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -531,39 +532,104 @@ class ConnectivityMapQuery(NetworkConfigQuery):
         peers_to_compare = self.config.peer_container.get_all_peers_group()
 
         ref_ip_blocks = IpBlock.disjoint_ip_blocks(self.config.get_referenced_ip_blocks(),
                                                    IpBlock.get_all_ips_block_peer_set())
-
+        connections = defaultdict(list)
+        peers = PeerSet()
         peers_to_compare |= ref_ip_blocks
-
-        conn_graph = ConnectivityGraph(peers_to_compare, self.config.allowed_labels, self.output_config,
-                                       self.config.type)
         for peer1 in peers_to_compare:
             for peer2 in peers_to_compare:
+                if self.is_in_subset(peer1):
+                    peers.add(peer1)
+                elif not self.is_in_subset(peer2):
+                    continue  # skipping pairs if none of them are in the given subset
                 if isinstance(peer1, IpBlock) and isinstance(peer2, IpBlock):
                     continue  # skipping pairs with ip-blocks for both src and dst
                 if peer1 == peer2:
-                    conn_graph.add_edge(peer1, peer2, ConnectionSet(True))  # cannot restrict pod's connection to itself
+                    # cannot restrict pod's connection to itself
+                    connections[ConnectionSet(True)].append((peer1, peer2))
                 else:
                     conns, _, _, _ = self.config.allowed_connections(peer1, peer2)
                     if conns:
                         if self.config.type == NetworkConfig.ConfigType.Istio and \
                                 self.output_config.connectivityFilterIstioEdges:
                             should_filter, modified_conns = self.filter_istio_edge(peer2, conns)
                             if not should_filter:
-                                conn_graph.add_edge(peer1, peer2, modified_conns)
+                                connections[modified_conns].append((peer1, peer2))
+                                # collect both peers, even if one of them is not in the subset
+                                peers.add(peer1)
+                                peers.add(peer2)
                         else:
-                            conn_graph.add_edge(peer1, peer2, conns)
+                            connections[conns].append((peer1, peer2))
+                            # collect both peers, even if one of them is not in the subset
+                            peers.add(peer1)
+                            peers.add(peer2)
 
         res = QueryAnswer(True)
         if self.output_config.outputFormat == 'dot':
+            conn_graph = ConnectivityGraph(peers, self.config.allowed_labels, self.output_config,
+                                           self.config.type)
+            conn_graph.add_edges(connections)
             res.output_explanation = conn_graph.get_connectivity_dot_format_str()
         else:
+            conn_graph = ConnectivityGraph(peers_to_compare, self.config.allowed_labels, self.output_config,
```

Comment:
Why in the case of textual reports we create the `ConnectivityGraph` using `peers_to_compare`? Wouldn't it be more efficient to create it using `peers`? What will we be missing?

Reply:
In the case of textual reports we aggregate the connections by label, so we need all the peers.